### PR TITLE
end to end testing pipeline

### PR DESCRIPTION
**Pull request by gumaerc**
# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/2033

# Description (What does it do?)
This PR adds a new pipeline definition, `TestPipelineDefiniton`. This pipeline is designed for running the end to end Playwright tests defined in `ocw-hugo-themes` against the output of the steps defined in `SitePipelineOnlineTasks`. The resulting pipeline will be automatically triggered on pushes to `ocw-hugo-themes` or `ocw-hugo-projects` based on configured branches and will alert us of any problems running the end to end tests against content that was built and deployed using the same steps we use to build and deploy any other site using pipelines generated by `ocw-studio`.

# How can this be tested?
 - Make sure your instance of `ocw-studio` is set up to publish websites according to the readme
 - Set the following environment variables:
```
OCW_WWW_TEST_SLUG=ocw-ci-test-www
OCW_COURSE_TEST_SLUG=ocw-ci-test-course
AWS_TEST_BUCKET_NAME=ocw-content-test
AWS_OFFLINE_TEST_BUCKET_NAME=ocw-content-offline-test
```
 - Download the JSON fixtures from this Google Drive folder: https://drive.google.com/drive/folders/1iqLw2C2Atk1_O94z7PpUC9X7JQe_IxSn
 - In your local Django admin, make sure you don't have existing sites that would conflict with the test courses. You can do this by going to http://localhost:8043/admin/websites/website/ and looking for `ocw-ci-test-course` and `ocw-ci-test-www`. If these websites already exist, delete them.
 - Place `test_websites.json` and `test_website_content.json` at the root of your `ocw-studio` folder
 - Run `docker compose exec web ./manage.py loaddata test_websites.json` and you should see `Installed 2 object(s) from 1 fixture(s)`
 - Run `docker compose exec web ./manage.py loaddata test_website_content.json` and you should see `Installed 43 object(s) from 1 fixture(s)`
 - Run `docker compose exec web ./manage.py reset_sync_states --filter "ocw-ci-test-course, ocw-ci-test-www" --skip_sync`
 - Run `docker compose exec web ./manage.py sync_website_to_backend --filter "ocw-ci-test-course, ocw-ci-test-www" --delete`
 - Browse to http://localhost:8043/sites/ocw-ci-test-www and publish the site live. Don't worry if the publish fails, the important part is that it gets pushed to the `release` branch in Github. To verify this browse to the repo in your test org and make sure that the site content is in the `release` branch.
 - Repeat the previous step for http://localhost:8043/sites/ocw-ci-test-course
 - Download the test course static resources from this Google Drive link: https://drive.google.com/drive/folders/14Hlid31Qy7Yy5V4OgHUwNleYUFuJ5BH2?usp=sharing
 - Open the Minio web UI at http://localhost:9001 and login with your credentials
 - Browse to the `ol-ocw-studio-app` bucket, then in the `courses` folder create a folder named `ocw-ci-test-course` if it doesn't already exist
 - Drag and drop the static resources we downloaded from Google Drive into the Minio folder we just created
 - Run `docker compose exec web ./manage.py upsert_theme_assets_pipeline --themes-branch cg/playwright-external-site`
 - Run `docker compose exec web ./manage.py upsert_e2e_test_pipeline --themes-branch cg/playwright-external-site`
 - Browse to your locally running Concourse instance at http://localhost:8080
 - Find the theme assets pipeline that we pushed up for `cg/playwright-external-site`, unpause it and trigger a build. The theme assets build has been altered as part of this PR to also copy the theme assets to the test buckets. Wait for it to finish.
 - Once the theme assets build has completed, locate the end to end testing pipeline and unpause it. The pipeline should eventually start automatically after about a minute. If it does not, check the inputs to the job and try refreshing them manually. Once the pipeline starts, it should complete without error.

# Additional Context
The offline build is currently not tested, but will be tested in the future and this pipeline will be updated to support it. Before this PR is merged, the new settings should be set in Heroku via a `salt-ops` PR. These settings include:

 - `OCW_WWW_TEST_SLUG`
 - `OCW_COURSE_TEST_SLUG`
 - `AWS_TEST_BUCKET_NAME`
 - `AWS_OFFLINE_TEST_BUCKET_NAME`

This `salt-ops` PR will not be created until https://github.com/mitodl/ol-infrastructure/issues/1945 is resolved and the new buckets that it will point at have been created.

### CODE REVIEW

**Comment by gumaerc** [comment]
@HussainTaj-arbisoft Thanks for your reviews. I took some of your suggestions and did the following:

 - `STATIC_API_BASE_URL_TEST` when running the pipeline locally needs to be the Docker IP of your nginx instance on the correct port. I made some changes to `content_sync/utils/get_common_pipeline_vars` to set the correct defaults, but for what it's worth that ends up being http://10.1.0.102:8046. In a browser you can access this at http://localhost:8046 because of port forwarding, but Concourse needs to access it through the Docker IP.
 - I altered the `docker-compose.yml` to create the test buckets
 - @ChristopherChudzicki pointed out that once this gets deployed to RC and production, the test sites will show up in the "new courses" section. This will cause the `ocw-www` build to fail, because the test sites aren't actually deployed to the publish bucket. I modified the API view sets to exclude the test sites based on `OCW_WWW_TEST_SLUG` and `OCW_COURSE_TEST_SLUG`.
 - I committed the fixtures to the repo under `test_site_fixtures`
 - I added a section to the readme about how to set up the pipeline locally

File: test_site_fixtures/test_website_content.json [hunk 1]
```diff
@@ -0,0 +1,1718 @@
...
+          "youtube_id": "",
+          "video_speakers": "",
+          "youtube_description": ""
+        },
+        "learning_resource_types": ["Activity Assignments"]
+      },
+      "is_page_content": true,
+      "filename": "example_jpg",
+      "dirpath": "content/resources",
+      "file": "courses/123-ocw-ci-test-course-fall-2022/example_jpg.jpg"
```
Comment (HussainTaj-arbisoft) [inline test_site_fixtures/test_website_content.json:44]: These need to be updated. https://github.com/mitodl/ocw-hugo-themes/pull/1280 changes expected urls from `.../123-ocw-ci-test-course-fall-2022/...` to `.../ocw-ci-test-course/...`.

The content generated from these fixtures fails tests.

**Review by HussainTaj-arbisoft — CHANGES_REQUESTED** [review]
@gumaerc, thank you for the updates. I got the pipeline up and running.

The tests are failing because of a URL mismatch in the fixtures and tests. I left a separate comment denoting an instance of the fixture that should be changed.

Here's a log of the `playwright-task` step, just in case: 
[playwright-task.log](https://github.com/mitodl/ocw-studio/files/13545939/playwright-task.log)

---

You might also be interested in this side note. I had my Docker setup with 8 GB memory. While running the e2e tests pipeline, my docker kept crashing. After I upgraded my docker's allocated memory to 10 GB, things started to work.

It's not in the scope of this ticket, but perhaps in the future, we might consider reducing the number of playwright workers for local development.

**Comment by gumaerc** [comment]
@HussainTaj-arbisoft I updated the url path of the offending files in the test site fixtures

**Review by HussainTaj-arbisoft — CHANGES_REQUESTED** [review]
@gumaerc, thanks for the updates. The fixtures seem correct now and some of the failing tests have started passing.

The test pipeline now fails for `file` fields.

```log
    Expected string: "https://10.1.0.102:8046/courses/ocw-ci-test-course/example_jpg.jpg"
    Received string: "https://10.1.0.102:8046/media/courses/ocw-ci-test-course/example_jpg.jpg"
```

I don't know where this `media` is coming from in the [synced content](https://github.com/hussaintaj-mitodl/ocw-ci-test-course/blob/main/content/resources/example_jpg.md). It might be either something wrong with my environment or this might be how local publishes work. Would you know anything about it?

---

PS: Left a couple of other comments.

**Review by HussainTaj-arbisoft — COMMENTED** [review]
Thank you, @gumaerc. I got the green lights 🎉 

I've left a few more questions and suggestions.

**Comment by gumaerc** [comment]
@HussainTaj-arbisoft I was able to address everything you noted above, but was not able to finish the unit tests tonight. I'll get that done first thing tomorrow morning.

**Review by HussainTaj-arbisoft — APPROVED** [review]
LGTM 👍 

I read the comment about dealing with WebsiteStarter fixtures later. That sounds okay to me.

**Pull request merged**